src/replit_river/server_transport.py

```diff
@@ -115,44 +137,26 @@ async def _get_or_create_session(
                     close_session_callback=self._delete_session,
                 )
             else:
-                old_session = self._sessions[to_id]
-                if old_session.session_id != session_id:
-                    logger.info(
-                        'Create new session with "%s" for session id %s'
-                        " and close old session %s",
-                        to_id,
-                        session_id,
-                        old_session.session_id,
-                    )
-                    session_to_close = old_session
-                    new_session = Session(
-                        transport_id,
-                        to_id,
-                        session_id,
-                        websocket,
-                        self._transport_options,
-                        self._is_server,
-                        self._handlers,
-                        close_session_callback=self._delete_session,
-                    )
-                else:
-                    # If the instance id is the same, we reuse the session and assign
-                    # a new websocket to it.
-                    logger.debug(
-                        'Reuse old session with "%s" using new ws: %s',
-                        to_id,
-                        websocket.id,
-                    )
-                    try:
-                        await old_session.replace_with_new_websocket(websocket)
-                        new_session = old_session
-                    except FailedSendingMessageException as e:
-                        raise e
+                # If the instance id is the same, we reuse the session and assign
+                # a new websocket to it.
+                logger.debug(
+                    'Reuse old session with "%s" using new ws: %s',
+                    to_id,
+                    websocket.id,
+                )
+                try:
+                    await old_session.replace_with_new_websocket(websocket)
+                    new_session = old_session
+                except FailedSendingMessageException as e:
+                    raise e
 
-            if session_to_close:
-                logger.info("Closing stale session %s", session_to_close.session_id)
-                await session_to_close.close()
+        if old_session and new_session != old_session:
+            logger.info("Closing stale session %s", old_session.session_id)
+            await old_session.close()
+
+        async with self._session_lock:
             self._set_session(new_session)
```

Comment:
for safety, one property that the old code had that this one doesn't is that there _could_ be two concurrent calls to this function (with the same id?) and then we would close the old session twice and leak one of the new sessions!

to prevent that what we can do is to rename `_set_session` to `_set_session_locked` (and add a docstring comment that it needs the `session_lock` to operate). and have `_set_session_locked` detect this mid-air collision and ... do something with the preexisting one (close it???)

with that out of the way, this relaxing of the locks' critical sections _seems_ to be safe, because the `Session` by itself is safe to create twice (as long as we close it, otherwise we leak tasks!!!)